### ISSUE TITLE
Replace `develop` branch with `develop-new`

### ISSUE THREAD

**Issue by kael-shipman** [cimasim89/flat-file-creator#33]
@cimasim89  Ok, it looks like there were a lot of problems that came up because of the rebasing, so I created a new `develop` branch to replace the current one. The new one is the old one, just rebased onto `master`. It has one small difference, and that's that it uses absolute versions rather than "compatible" versions for dependencies. (I know you said you wanted to use absolute versions anyway, so I'm not sure how we ended up with compatible versions in the real `develop` branch, but this fixes that.)

**If you approve of this, please _replace_ the current `develop` branch with `develop-new`.** If you don't approve, then please conduct the merge of the current `develop` branch into `master` as you wish and rectify branches for new development.

On a more procedural note, I'd like to humbly suggest we try a different branch management strategy. I believe you had originally indicated that you prefer gitflow. However, while gitflow works well for services, I think libraries can benefit from a flow that preserves their versioning better.

Here's my suggestion:

* **Rename `master` to `v2.x`.** You can also retroactively go create a `v1.x` branch if you'd like by checking out the [1.1.0 tag](https://github.com/cimasim89/flat-file-creator/releases/tag/1.1.0).
* **Set the default branch to `v2.x`.**
* **New features are branched off of the current default branch (in this case `v2.x`).** That way, if you need to release the new feature before other features are ready, you can always do so without accidentally including other development that's not ready yet.
* **Feature branches are _squash-merged_ into the main version branch,** so that the history of the main version branches remains clean and readable.
* **Alternatively, if desired, branches of the form `v2.1.x` may be branched off of the main branch in preparation for the release of that version.** As features become ready, you can safely merge them into this release branch, which will only ever contain code that is presumed ready for production.
* **Bugfixes are branched from the version branch in which the bug first appears,** then squash-merged into all version branches to which they apply.

The result of this flow is that you have major version branches, each containing discrete feature/bugfix commits and tags along the way indicating which features or fixes were applied to which releases. Additionally, you can easily go back and continue to support development on previous versions in the event of bugfixes or features that are easy to back-port.

Anyway, I know how personal branching strategy is, and I know this is your project, not mine, so feel free to disregard. Either way, though, please do provide clarity on when you'd like to rebase vs squash-merge vs regular merge so that we don't run into more problems.

**Comment by cimasim89** [comment]
Hi Kael, currently so develop-new is working properly right? I'll just try to fix develop and ask you to give me feedback on the proposal as soon as I'm done. For the different branch management, we can switch to this kind of strategy, the main thing is that I didn't expect the library to have this growth and improvement.

**Comment by kael-shipman** [comment]
Ok, sounds good. Lemme know if/when you'd like my input.

As far as I'm concerned, once we get all the branches merged and rectified, we're done and ready for release, and I won't have any more updates for a while (other than possible bugfixes). This has been fun!

**Comment by cimasim89** [comment]
> Anyway, I know how personal branching strategy is, and I know this is your project, not mine, so feel free to disregard. 

As far as branching strategy is concerned, I'm open to using new techniques, only the tight schedule often doesn't give me a chance to explore or appraise certain types of improvements, so I tend to stick with what I've already used in other projects. Anyway, as always you have been very precise in proposing this different strategy, and I seem to understand its strengths, even if in these cases I realize that I have some gaps in the deep use of the git features and I'm a little bit relaxed on the use I normally make of them. Now I'll try to follow your advice about branch management, it's very likely that I'll find myself bothering you with some questions.

> Either way, though, please do provide clarity on when you'd like to rebase vs squash-merge vs regular merge so that we don't run into more problems.

We can also decide now which is the best or the most adherent way to this branch management.

**Comment by kael-shipman** [comment]
> Now I'll try to follow your advice about branch management, it's very likely that I'll find myself bothering you with some questions.

No problem at all! Strategy discussions are really interesting to me because they represent the act of developing and advancing tech culture together and I think that's a fascinating process. I know I don't have any "answers" - just ideas - so it's fun to hash out together the best way to manage this stuff.

I think the main pain point that I've been dealing with lately is keeping history concise, but _related._ The thing you and I ran into here is that the rebases we did caused `master` and `develop` to become unrelated, so branching off of `master` and merging into `develop` caused a duplication of a bunch of commits that were technically different but applied the same changes. In this case it was rebases, but the same thing can happen with squash merges.

For me, the easiest way to solve that is to establish a rule around when we should rebase, when we should squash-merge and when we should regular-merge.

To clarify my proposal above: I believe we should **rebase when pulling feature branches** down from origin, **squash-merge feature branches into main branches**, then always **regular-merge main branches together** when applicable.

To be clear about the definitions I'm using for this, in our case right now, the "main" branches are `develop` and `master`. That means if we branch `feat-abc` from `master`, we will always `git pull --rebase` when pulling changes; then we will squash-merge `feat-abc` into `develop` at the end of the PR process; then regular-merge `develop` into `master` when we're ready to release an official version.

Does that sound like a decent ruleset? Lemme know if you see problems or have other preferences.

(Note that this is _separate_ from the discussion of whether or not we move from `master`/`develop` as main branches to `v1.x`/`v2.x`/`v3.x` as main branches. In the latter case, if we still want to maintain the idea of a separate "release branch" comparable to the current `develop` branch, we can use a naming convention like `v1.2.3-rc`, which would be the release candidate branch for the `v1.2.3` release.)
